File: sim/src/chap1.zig
const std = @import("std");
const testing = std.testing;
const assert = std.debug.assert;
const expect = std.testing.expect;
const eql = std.mem.eql;

pub fn mean(comptime T: type, data: []const T) f64 {
    assert(data.len > 0); // not accept empty data

    var sum: f64 = 0;

    switch (T) {
        i16, i32, i64 => for (data) |value| {
            sum += @as(f32, @floatFromInt(value));
        },

        f16, f32, f64 => for (data) |value| {
            sum += value;
        },

        else => @compileError("Not support this type!"),
    }

    return sum / @as(f32, @floatFromInt(data.len));
}

pub fn trimmedMean(comptime T: type, data: []const T, trim: usize) f64 {
    assert(data.len > trim * 2); // the trimming not over data length

    //Make a copy
    const allocator = std.heap.page_allocator;
    const copy = allocator.dupe(T, data) catch {
        @panic("Can't make a copy");
    };
    defer allocator.free(copy);

    std.mem.sort(T, copy, {}, comptime std.sort.asc(T));

    const trimmed_data = copy[trim..(copy.len - trim)];
    assert(trimmed_data.len == (copy.len - 2 * trim));

    return mean(T, trimmed_data);
}

pub fn weightedMean(comptime T: type, data: []const T, weights: []const f16) f64 {
    assert(data.len == weights.len); // each weight must belong to a entry

    var weighted: f64 = 0;
    var sum_weights: f64 = 0;

    switch (T) {
        i16, i32, i64 => for (0..(data.len)) |i| {
            weighted += @as(f64, @floatFromInt(data[i])) * weights[i];
            sum_weights += weights[i];
        },
        f16, f32, f64 => for (0..(data.len)) |i| {
            weighted += data[i] * weights[i];
            sum_weights += weights[i];
        },

        else => @compileError("Not support this type!"),
    }

    return weighted / sum_weights;
}

pub fn median(comptime T: type, data: []const T) f64 {
    assert(data.len > 1);

    //Make a copy
    const allocator = std.heap.page_allocator;
    const copy = allocator.dupe(T, data) catch {
        @panic("Can't make a copy");
    };
    defer allocator.free(copy);

    std.mem.sort(T, copy, {}, comptime std.sort.asc(T));

    switch (T) {
        i16, i32, i64 => if (data.len % 2 == 1) {
            return @as(f64, @floatFromInt(data[data.len / 2]));
        } else {
            return @as(f64, @floatFromInt(data[(data.len / 2) - 1] + data[data.len / 2])) / 2;
        },

        f16, f32, f64 => if (data.len % 2 == 1) {
            return data[data.len / 2];
        } else {
            return (data[(data.len / 2) - 1] + data[data.len / 2]) / 2;
        },

        else => @compileError("Not support this type!"),
    }
}

test "test mean 1" {
    const data = [_]f32{ 1, 2, 3 };
    try testing.expect(mean(f32, &data) == 2);
}

test "test mean 2" {
    const data = [_]f32{ 1, 2 };
    try testing.expect(mean(f32, &data) == 1.5);
}

test "test mean 3" {
    const data = [_]i32{ 1, 2 };
    try testing.expect(mean(i32, &data) == 1.5);
}

test "test trimmed mean 1" {
    var data = [_]i32{ 0, 1, 2, 0 };
    try testing.expect(trimmedMean(i32, &data, 1) == 0.5);
}

test "test trimmed mean 2" {
    var data = [_]f16{ 0, 1.5, 2, 0 };
    try testing.expect(trimmedMean(@TypeOf(data[0]), &data, 1) == 0.75);
}

test "test trimmed mean 3" {
    var data = [_]f16{ 0, 1, 2, 1, 3 };
    try testing.expect(trimmedMean(@TypeOf(data[0]), &data, 2) == 1);
}

test "test weighted mean 1" {
    const data = [_]f16{ 2, 4 };
    const weights = [_]f16{ 0.5, 0.5 };
    try testing.expect(weightedMean(@TypeOf(data[0]), &data, &weights) == 3.0);
}

test "test weighted mean 2" {
    const data = [_]i16{ 2, 4 };
    const weights = [_]f16{ 0.5, 0.5 };
    try testing.expect(weightedMean(@TypeOf(data[0]), &data, &weights) == 3.0);
}

test "test weighted mean 3" {
    const data = [_]f32{ 2, 4 };
    const weights = [_]f16{ 0.5, 0.5 };
    try testing.expect(weightedMean(@TypeOf(data[0]), &data, &weights) == 3.0);
}
test "test median 1" {
    const data = [_]f32{ 2, 4 };
    try testing.expect(median(@TypeOf(data[0]), &data) == 3);
}
test "test median 2" {
    const data = [_]i32{ 2, 3 };
    try testing.expect(median(@TypeOf(data[0]), &data) == 2.5);
}

File: sim/C/src/chap1.c
#include <assert.h>
#include <math.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "mylib/slice.c"

typedef Slice Data;

double const *get_ele(Data const data, size_t const index) {
  return (double *)get_item(data, index);
}
Data copy_data(Data const data) { return copy_slice(data); }
void qsort_data(Data const data) { qsort_slice(data); }
Data make_data(Data const data, size_t low, size_t high) {
  return make_slice(data, low, high);
}
Data sorted_data(Data const data) { return sorted_slice(data); }

void free_data(Data const data) { free_slice(data); }

double mean_double(Data const data) {
  assert(data.len >= 1);

  double sum = 0;

  for (size_t i = 0; i < data.len; ++i) {
    sum += *get_ele(data, i);
    assert(!isinf(sum) && "check overflow");
  }

  return sum / data.len;
}

double median_double(Data const data) {
  assert(data.len >= 1);

  Data const sorted_copy = sorted_slice(data);

  double median;
  if (data.len % 2 == 1) {
    median = *get_ele(sorted_copy, sorted_copy.len / 2);
  } else {
    median = (*get_ele(sorted_copy, sorted_copy.len / 2) +
              *get_ele(sorted_copy, sorted_copy.len / 2 - 1)) /
             2;
  }

  free_data(sorted_copy);

  assert(!isinf(median) && "check overflow");
  return median;
}

double trim_mean(Data const data, size_t const trim) {
  assert(data.len > trim * 2 && "the trimming not over data length");

  Data const sorted_copy = sorted_slice(data);

  // Data const trimmed_data = {.pointer = sorted_copy + trim,
  //                             .len = len - 2 * trim,
  //                             .item_size = sizeof(double)};
  //
  Data const trimmed_data =
      make_slice(sorted_copy, trim, sorted_copy.len - trim);

  assert(trimmed_data.len == sorted_copy.len - 2 * trim &&
         "check valid trimmed");

  double const mean = mean_double(trimmed_data);

  free_data(sorted_copy);

  return mean;
}

double weighted_mean(Data const data, Slice const weights) {

  assert(data.len == weights.len && "each weight must belong to an entry");

  double weighted = 0.0;
  double sum_weights = 0.0;

  for (size_t i = 0; i < data.len; ++i) {
    weighted += (*get_ele(data, i)) * (*get_ele(weights, i));
    assert(!isinf(weighted) && "overflow check");
    sum_weights += *get_ele(weights, i);
    assert(!isinf(sum_weights) && "overflow check");
  }

  assert(sum_weights != 0 && "total weight can not be zero");
  return weighted / sum_weights;
}

double var(Data const data, size_t ddof) {
  assert(data.len > 0);
  assert(ddof < data.len &&
         "degree of freedom delta is always < length of data");

  double sum_squares = 0;
  double const mean = mean_double(data);

  for (size_t i = 0; i < data.len; ++i) {
    sum_squares += pow(*get_ele(data, i) - mean, 2);
    assert(!isinf(sum_squares) && "overflow check");
  }

  return sum_squares / (data.len - ddof);
}

double std(Data const data, size_t ddof) {
  assert(data.len > 0);
  assert(ddof < data.len &&
         "degree of freedom delta is always < length of data");

  return sqrt(var(data, ddof));
}

bool is_percent(double const percent) {
  if (percent > 0.0 && percent < 1.0) {
    return true;
  }
  return false;
}

// ref: https://numpy.org/doc/2.1/reference/generated/numpy.quantile.html
double quantile(Data const data, double const q) {
  assert(data.len > 0);
  assert(is_percent(q));

  Data sorted = sorted_data(data);

  double j = -1; // -1 is dummy number to check
  double const g = modf(q * (data.len - 1), &j);
  assert(j >= 0);

  double const result =
      (1 - g) * (*get_ele(sorted, j)) + g * (*get_ele(sorted, j + 1));

  free_data(sorted);
  return result;
}

// this main function for static analysis and quick test

// int main(void) {
//
//   double const arr[] = {1, 2, 3, 4};
//
//   Data const data = {
//       .pointer = (char *)arr, .len = 4, .item_size = sizeof(double)};
//
//   printf("%lf\n", quantile(data, 0.25));
//   return EXIT_SUCCESS;
// }
